nodejs/awsx/apigateway/apikey.ts

```diff
@@ -0,0 +1,147 @@
+// Copyright 2016-2018, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// These APIs are currently experimental and may change.
+
+import * as aws from "@pulumi/aws";
+import * as api from "./api";
+
+export type APIKeySource = "HEADER" | "AUTHORIZER";
+
+/**
+ * Input properties for creating a UsagePlan and associated API Keys.
+ */
+export interface APIKeyArgs {
+
+    /**
+     * Define the apis you would like to associate the usage plan with. This can be used in place of
+     * defining the apiStages defined in the [usagePlan]. You cannot define [apis] and [usagePlan.apiStages].
+     */
+    apis?: api.API[];
+
+    /**
+     * Define the usage plan to create. If an existing Usage Plan is passed in, then the API Keys
+     * will be associated with the usage plan and no [apis] can be added.
+     */
+    usagePlan?: aws.apigateway.UsagePlanArgs | aws.apigateway.UsagePlan;
```

Comment:
it feels like this is the most important thing, and should potentially be first.  i.e. if there is an [aws.apigwateway.UsagePlan] provdied, then we'll just use that.  Then, if we don't have that, we'll combine the UsagePlanArgs with [apis] to produce the usage plan.

Note: i had to read through a few time to figure out the set of combinations of all this stuff which are valid.  it might be worth enumerating.  i.e.:

1. You can provided [apis] and nothing else, in which case: ...
2. You can provider [apis] and [UsagePlanArgs], in which case...
3. You can provide a pre-existing UsagePlan.  In this case, [apis] cannot be provided.
etc. etc.

Reply:
Also proably worth mentioning what hapens if both are not provided.

Reply:
I think I've addressed this but please let me know if its still not clear!